### PR TITLE
Overdue facility filtering

### PR DESCRIPTION
**Pull request by timcheadle**
*Stories* 
- https://www.pivotaltracker.com/story/show/164033363
- https://www.pivotaltracker.com/story/show/164131252
- https://www.pivotaltracker.com/story/show/164107023
- https://www.pivotaltracker.com/story/show/164132709

*Notes*
- Allow admins to filter the overdue list by facility
- Allow admins to specify how many records per page they want
- Order facility names alphanumerically
- Refactor: create shared partials for
   - pagination/filtering by facility
   - patient followup card
- Add controller/feature specs for pagination/filtering by facility
- Bugfixes
   - highlight Overdue Appointments tab when selected
   - indicate when no phone number available in 48 hr list

*Breakdown*
- dependency updates: 2 files (gemfile/lockfile)
- pagination mixin: 1 file
- controller updates: 2 files (appointments/patients)
- shared partials: 2 files (pagination/followup-card)
- view updates: 3 files (appointments/patients + small fix in layout)
- specs: 5 files (feature/controller)

Total: 2 + 1 + 2 + 2 + 3 + 5 = 15 files

### CODE REVIEW

**Comment by divs1210** [comment]
- Overdue Page
  - All Facilities
    overdue-all
  - Selected facility
     overdue-facility
- 48 Hr List
  - All Facilities
     48-all
  - Selected Facility
     48-facility

File: app/controllers/appointments_controller.rb [hunk 1]
```diff
@@ -3,16 +3,18 @@ class AppointmentsController < AdminController
...
     authorize Appointment, :index?
-    @appointments = policy_scope(Appointment)
-                      .overdue
-                      .order(scheduled_date: :asc)
-                      .page(params[:page]).per(10)
-  end
 
-  def edit
-  end
+    @facility_id = params[:facility_id].present? ? params[:facility_id] : 'All'
```
Comment (govindkrjoshi) [inline app/controllers/appointments_controller.rb]:
```suggestion
    @facilities = params[:facility_id].present? ? Facility.where(id: params[:facility_id]) : policy_scope(Facility)
```

File: app/controllers/appointments_controller.rb [hunk 2]
```diff
@@ -3,16 +3,18 @@ class AppointmentsController < AdminController
...
+    @per_page = params[:per_page] || 10
 
-  def cancel
+    appointments_to_show = policy_scope(Appointment)
+                             .overdue
+                             .where_or_all(:facility_id, @facility_id)
```
Comment (govindkrjoshi) [inline app/controllers/appointments_controller.rb]:
```suggestion
                             .where(facility, facilities)
```

File: app/controllers/appointments_controller.rb (hunk 1 above)
Comment (govindkrjoshi) [inline app/controllers/appointments_controller.rb]:
```suggestion
    @selected_facility_id = params[:facility_id].present? ? params[:facility_id] : 'All'
    @facilites = @selected_facility_id == 'all' ? policy_scope(Facility.all) : policy_scope(Facility.where(id: @selected_facility_id))
```

File: app/controllers/appointments_controller.rb [hunk 3]
```diff
@@ -3,16 +3,18 @@ class AppointmentsController < AdminController
...
-    @appointments = policy_scope(Appointment)
-                      .overdue
-                      .order(scheduled_date: :asc)
-                      .page(params[:page]).per(10)
-  end
 
-  def edit
-  end
+    @facility_id = params[:facility_id].present? ? params[:facility_id] : 'All'
+    @per_page = params[:per_page] || 10
```
Comment (govindkrjoshi) [inline app/controllers/appointments_controller.rb]:
```suggestion
    @per_page = params[:per_page] == 'All' ? appointments_to_show.size : params[:per_page].to_i
```

File: app/controllers/appointments_controller.rb [hunk 4]
```diff
@@ -3,16 +3,18 @@ class AppointmentsController < AdminController
...
 
-  def cancel
+    appointments_to_show = policy_scope(Appointment)
+                             .overdue
+                             .where_or_all(:facility_id, @facility_id)
+
+    @appointments = appointments_to_show
+                      .order(scheduled_date: :asc)
+                      .page(params[:page])
+                      .per(@per_page == 'All' ? appointments_to_show.count : @per_page.to_i)
```
Comment (govindkrjoshi) [inline app/controllers/appointments_controller.rb]:
```suggestion
                      .per(@per_page || DEFAULT_PAGE_SIZE)
```

File: app/controllers/appointments_controller.rb [hunk 5]
```diff
@@ -3,16 +3,18 @@ class AppointmentsController < AdminController
```
Comment (govindkrjoshi) [inline app/controllers/appointments_controller.rb]:
```suggestion
  DEFAULT_PAGE_SIZE = 10
```

**Comment by divs1210** [comment]
@govind CodeClimate still complains about somethings. How should we refactor to address them?

File: app/controllers/patients_controller.rb [hunk 6]
```diff
@@ -1,12 +1,23 @@
...
+  DEFAULT_PAGE_SIZE = 20
+
   def index
     authorize Patient, :index?
-    @patients = policy_scope(Patient)
-                               .not_contacted
-                               .order(device_created_at: :asc)
-                               .page(params[:page]).per(10)
+
+    @facility_id = params[:facility_id].present? ? params[:facility_id] : 'All'
```
Comment (govindkrjoshi) [inline app/controllers/patients_controller.rb]:
```suggestion
    @facility_id = params[:facility_id] : 'All'
```

File: app/controllers/patients_controller.rb [hunk 7]
```diff
@@ -1,12 +1,23 @@
...
+    selected_facilities = @facility_id == 'All' ? policy_scope(Facility.all) : policy_scope(Facility.where(id: @facility_id))
+    patients_to_show = policy_scope(Patient)
+                         .not_contacted
+                         .where(registration_facility: selected_facilities)
+
+    @per_page = params[:per_page].present? ? params[:per_page] : DEFAULT_PAGE_SIZE
```
Comment (govindkrjoshi) [inline app/controllers/patients_controller.rb]:
```suggestion
    @per_page = params[:per_page] : DEFAULT_PAGE_SIZE
```

File: app/controllers/patients_controller.rb [hunk 8]
```diff
@@ -1,12 +1,23 @@
...
+    per_page_count = @per_page == 'All' ? patients_to_show.size : @per_page.to_i
```
Comment (govindkrjoshi) [inline app/controllers/patients_controller.rb]: You can set this before assigning `patients_to_show`, in which case the assignment becomes redundant.

File: app/controllers/patients_controller.rb [hunk 9]
```diff
@@ -1,12 +1,23 @@
...
+    @patients = patients_to_show
+                  .order(device_created_at: :asc)
+                  .page(params[:page])
```
Comment (govindkrjoshi) [inline app/controllers/patients_controller.rb]: What happens when params[:page] is `nil`?

File: app/views/patients/index.html.erb [hunk 10]
```diff
@@ -1,58 +1,92 @@
 <h1>48 Hour Patient Followups</h1>
 
-<% @patients.group_by(&:registration_facility).each do |facility, patients| %>
+<div class="row">
+  <div class="col-md-6">
+    <%= bootstrap_form_with(url: patients_path, method: :get, layout: :horizontal, class: "mt-4") do |form| %>
```
Comment (govindkrjoshi) [inline app/views/patients/index.html.erb]: We are using this component in 2 places. We can extract this out as a partial.

File: app/views/appointments/index.html.erb [hunk 11]
```diff
@@ -1,72 +1,100 @@
...
-                <% end %>
-              </div>
-            </div>
+<div id="overdue-patients">
+  <%
+    appointments_with_patients = @appointments.select { |a| a.patient.present? }
+    if appointments_with_patients.present?
+  %>
+    <% appointments_with_patients.each do |appointment| %>
+      <section class="overdue-patient">
```
Comment (govindkrjoshi) [inline app/views/appointments/index.html.erb]: We can also extract the overdue patient component to a partial.

File: app/controllers/patients_controller.rb (hunk 6 above)
Comment (divs1210) [reply app/controllers/patients_controller.rb]: done

File: app/controllers/patients_controller.rb (hunk 7 above)
Comment (divs1210) [reply app/controllers/patients_controller.rb]: done

File: app/controllers/patients_controller.rb (hunk 6 above)
Comment (divs1210) [reply app/controllers/patients_controller.rb]: done

File: app/controllers/patients_controller.rb (hunk 7 above)
Comment (divs1210) [reply app/controllers/patients_controller.rb]: done

File: app/controllers/patients_controller.rb (hunk 8 above)
Comment (divs1210) [reply app/controllers/patients_controller.rb]: clarified

File: app/controllers/patients_controller.rb (hunk 9 above)
Comment (divs1210) [reply app/controllers/patients_controller.rb]: the `page` method and `params[:page]` are both provided by `kaminari`, and I'm not seeing a case where this happens.

File: app/views/appointments/index.html.erb (hunk 11 above)
Comment (divs1210) [reply app/views/appointments/index.html.erb]: done

File: app/views/patients/index.html.erb (hunk 10 above)
Comment (divs1210) [reply app/views/patients/index.html.erb]: done

File: app/views/appointments/index.html.erb [hunk 12]
```diff
@@ -1,4 +1,4 @@
-<h1>Overdue patients</h1>
+<h1>Overdue Patients</h1>
```
Comment (timcheadle) [inline app/views/appointments/index.html.erb]: We use sentence case throughout this app; please don't change this.

**Comment by divs1210** [comment]
> This PR is massive for a small change. Let's stop doing code refactors when what we're really trying to change is adding a feature. (I'm totally guilty of this and will stop.)

@timcheadle I totally agree. I would prefer doing refactors as separate stories so the feature PRs are clearer. /cc @govindkrjoshi 

> This is really hard to review and exposes us to significant risk as it's changing a LOT.
> 
> To be honest I'm not sure how to even go about reviewing this, as the description doesn't even include a summary of changes now.

I've updated the summary to be more accurate, and I believe it now covers all the major changes.

File: app/views/appointments/index.html.erb (hunk 12 above)
Comment (divs1210) [reply app/views/appointments/index.html.erb]: will fix this

**Comment by timcheadle** [comment]
I refactored and fixed some things. Since this is my own PR I can't approve it, but I'm approving it here :)

**Pull request merged**